tools/track_participation.py

```diff
@@ -0,0 +1,215 @@
+
+config = json.load(open("tools/track_participation_config.json"))
+
+LECTURE_DATES_TO_NUMBER = config['LECTURE_DATES_TO_NUMBER']
+LECTURE_DATES_TO_START_TIME = config['LECTURE_DATES_TO_START_TIME']
+COMMENTING_DURATION_HOURS = config['COMMENTING_DURATION_HOURS']
+LECTURE_TIMEZONE = ZoneInfo(config['LECTURE_TIMEZONE'])
+
+
+def main():
+    handle_args(sys.argv[1:])
+
+    repo, issue = get_repo_and_issue()
+
+    participation = get_participation(issue)
+
+    print_content = ""
+
+    if PRINT_IN_MARKDOWN:
+        print_content = get_participation_markdown(participation)
+    else:
+        print_content = get_participation_text(participation)    
+
+    if PUBLISH:
+        update_issue_description(participation, issue)
+        print("Updated issue description\n")
+
+    print(print_content)
+
+
+def get_repo_and_issue():
+    """
+    Attempts to fetch the GitHub repository and issue using environment variables.
+    Exits on failure.
+    """
+    try:
+        github = Github(GITHUB_TOKEN)
+        repo = github.get_repo(REPO_FULLNAME)
+        issue = repo.get_issue(number=int(ISSUE_NUMBER))
+        return repo, issue
+    except GithubException as e:
+        logging.error(f"GitHub API error: {str(e)}")
+        sys.exit(1)
+
+
+def get_participation(tracking_issue):
+    """
+    Gets participation by checking each comment on the tracker issue.
+    Adds the lecture to the comment author if comment exists. Except if:
+        comment is from collaborator
+        comment is made outside of allowed time.
+    """
+    participation = {}
+
+    try:
+        collaborators = {collaborator.login for collaborator in tracking_issue.repository.get_collaborators()}
+    except GithubException as e:
+        logging.error(f"GitHub API error: {str(e)}")
+        sys.exit(1)
+
+    for comment in tracking_issue.get_comments():
+        author = comment.user.login
+        comment_time = comment.created_at.astimezone(LECTURE_TIMEZONE)
+
+        # Ignore collaborators' comments
+        if author in collaborators:
+            continue
+
+        if is_valid_lecture_time(comment_time):
+            lecture_date = comment_time.strftime("%Y-%m-%d")
+            if author not in participation:
+                participation[author] = {lecture_date}
+            else:
+                participation[author].add(lecture_date)
+
+    return participation
+
+
+def is_valid_lecture_time(comment_time):
+    """
+    Checks if the comment was made on a valid day and time.
+    """
+    lecture_date_str = comment_time.strftime("%Y-%m-%d")
+
+    if lecture_date_str not in LECTURE_DATES_TO_START_TIME:
+        # comment not made on a lecture day
+        return False
+
+    lecture_start_hour = LECTURE_DATES_TO_START_TIME[lecture_date_str]
+
+    allowed_period_start = comment_time.replace(hour=lecture_start_hour, minute=0, second=0, microsecond=0)
+    allowed_period_end = allowed_period_start + timedelta(hours=COMMENTING_DURATION_HOURS)
```

Comment:
Sorry if I being too nitpicky here, but I would request you to have two variables here instead of one `COMMENT_DURATION_HOURS`. One which is the actual duration and the other which is the leeway. This would be my personal opinion in terms of code readability/understandability.

Reply:
Agreed! addressed in ef44d53